docs/blocks.py

```diff
@@ -0,0 +1,168 @@
+        if code.startswith("```python"):
+            code = code[9:]
+            code = code[:-3]
+        code = textwrap.dedent(code)
+
+        app_width: str = self.options["app_width"]
+        mode: str = self.options["mode"]
+        url = create_marimo_app_url(
+            code=create_marimo_app_code(code=code, app_width=app_width),
+            mode=mode,
+        )
+        self._create_iframe(block, url)
+
+
+class MarimoEmbedFileBlock(BaseMarimoBlock):
+    NAME: str = "marimo-embed-file"
+    OPTIONS: dict[str, list[str | Any]] = {
+        **BaseMarimoBlock.OPTIONS,
+        "filepath": ["", type_string],
+        "show_source": ["true", type_string_in(["true", "false"])],
+    }
+
+    def on_end(self, block: etree.Element) -> None:
+        filepath = self.options["filepath"]
+        if not filepath:
+            raise ValueError("File path must be provided")
+
+        # Read from project root
+        try:
+            with open(filepath) as f:
+                code = f.read()
+        except FileNotFoundError:
+            raise ValueError(f"File not found: {filepath}")
+
+        mode: str = self.options["mode"]
+        url = create_marimo_app_url(code=code, mode=mode)
+        self._create_iframe(block, url)
+
+        # Add source code section if enabled
+        show_source: str = self.options.get("show_source", "true")
+        if show_source == "true":
+            details = etree.SubElement(block, "details")
+            summary = etree.SubElement(details, "summary")
+            summary.text = f"Source code for `{filepath}`"
+
+            # TODO: figure out syntax highlighting
+            # md_text = f"\n\n```python\n{code}\n```\n\n"
+            # result = self.md.htmlStash.store(self.md.convert(md_text))
+            # container.text = result
+
+            copy_paste_container = etree.SubElement(details, "p")
+            copy_paste_container.text = (
+                "Tip: paste this code into an empty cell, and the marimo editor will "
+                "create cells for you"
+            )
+
+            code_container = etree.SubElement(details, "pre")
+            code_container.set("class", "marimo-source-code")
+            code_block = etree.SubElement(code_container, "code")
+            code_block.set("class", "language-python")
+            code_block.text = code
+
+
+def create_marimo_app_code(
+    *,
+    code: str,
+    app_width: str = "wide",
+) -> str:
+    header = "\n".join(
+        [
+            "import marimo",
+            f'app = marimo.App(width="{app_width}")',
+            "",
+        ]
+    ) + "\n".join(
+        [
+            "",
+            "@app.cell",
+            "def __():",
+            "    import marimo as mo",
+            "    return",
+        ]
+    )
+    return header + code
+
+
+def create_marimo_app_url(code: str, mode: str = "read") -> str:
+    from lzstring2 import LZString
+
+    encoded_code = LZString.compress_to_encoded_URI_component(code)
```

Comment:
This is the main difference with their version, I reimplemented the compression function they use to generate the URL in Python.